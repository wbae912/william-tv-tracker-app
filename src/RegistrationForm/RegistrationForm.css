.register-section {
  box-sizing: border-box;
  font-family: 'Montserrat', sans-serif;
  color: white;
  text-align: center;
  width: 90%;
  margin-right: auto;
  margin-left: auto;
  margin-top: 5%;
}

.register-h2 {
  margin-top: 0;
  margin-top: 4%
}

.register-form-div {
  display: flex;
  justify-content: center;
}

.register-label {
  display: inline-block;
  width: 150px;
}

.register-input {
  margin-left: 2%;
  margin-right: 2%;
}

.buttons-flex {
  display: flex;
  margin-top: 2%;
}

.buttons-flex > .back-button-div, .signup-button-div {
  flex: 1
}

.signup-button-div {
  display: flex;
  justify-content: flex-end;
}

.back-button-div {
  display: flex;
  justify-content: flex-start;
}

.signup-button, .back-button {
  background-color: transparent;
  color: white;
  border-radius: 10px;
  border: 1.5px solid white;
  width: 80px;
  height: 30px;
  font-size: 14px;
  transition-duration: 0.4s;
}

.signup-button {
  margin-right: 1%;
}

.back-button {
  margin-left: 1%;
}

.signup-button:hover, .back-button:hover {
  background-color: white;
  color: black;
}