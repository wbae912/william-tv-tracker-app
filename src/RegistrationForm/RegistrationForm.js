import React from 'react';
import FormError from '../FormError/FormError';
import './RegistrationForm.css';

export default class RegistrationForm extends React.Component {
  constructor(props) {
    super(props)
  
    this.state = {
       error: null,
       full_name: '',
       user_name: '',
       password: '',
       fullNameTouch: {
         touched: false
       },
       userNameTouch: {
         touched: false
       },
       passwordTouch: {
        touched: false
      }
    }
  }
  
  handleChange = e => {
    this.setState({
    [e.target.name]: e.target.value
    })
  }

  handleFullNameTouch = (e) => {
    this.setState({
      fullNameTouch: {touched: true}
    })
  }

  handleUserNameTouch = (e) => {
    this.setState({
      userNameTouch: {touched: true}
    })
  }

  handlePasswordTouch = (e) => {
    this.setState({
      passwordTouch: {touched: true}
    })
  }

  handleSubmit = e => {
    e.preventDefault();
    const credentials = {
      full_name: this.state.full_name,
      user_name: this.state.user_name,
      password: this.state.password
    }

    fetch('http://localhost:8000/api/users', {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(credentials)
    })
    .then(res => {
      if(!res.ok) {
        return res.json().then(err => Promise.reject(err));
      }
      return res.json();
    })
    .then(() => {
      this.props.history.push('/login')
    })
    .catch(res => {
      this.setState({
        error: res.error
      })
    })
  }

  validateFullName = () => {
    const fullName = this.state.full_name;
    if(fullName.length === 0) {
      return 'Please enter your full name';
    }
  }

  validateUserName = () => {
    const username = this.state.user_name;
    if(username.length === 0) {
      return 'Please enter a username';
    }
  }

  validatePassword = () => {
    const password = this.state.password;
    const REGEX = /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])/
    if(password.length === 0) {
        return 'Please enter a password';
    } else if(password.length < 8 || password.length > 72) {
        return 'Password must be between 8 and 72 characters';
    }
    if (password.startsWith(' ') || password.endsWith(' ')) {
      return 'Password must not start or end with empty spaces';
    }
    if (!REGEX.test(password)) {
      return 'Password must contain at least one upper case, lower case, and number character';
    }
  }

  render() {
    let fullNameError = this.validateFullName();
    let passwordError = this.validatePassword();
    let userNameError = this.validateUserName();

    return (
      <section className="register-section">
        <div role="alert" className="error-div">
          {this.state.error && <p className='red'>{this.state.error}</p>}
        </div>
        <h2 className="register-h2">Start Today!</h2>
        <form className="register-form" onSubmit={e => this.handleSubmit(e)}>
          <div className="register-form-div">
            <label htmlFor="full_name" className="register-label">Full Name</label>
            <input type="text" className="register-input" id="full_name" name="full_name" placeholder="Bob Smith" required
              aria-required="true" aria-label="Full name" aria-invalid="true"
              onChange={e => {this.handleChange(e); this.handleFullNameTouch(e)}}/>
          </div>
          {this.state.fullNameTouch.touched && <FormError message={fullNameError} />}
          <div className="register-form-div">
            <label htmlFor="user_name" className="register-label">Username</label>
            <input type="text" className="register-input" id="user_name" name="user_name" placeholder="username" required
              aria-required="true" aria-label="username" aria-invalid="true"
              onChange={e => {this.handleChange(e); this.handleUserNameTouch(e)}}/>
          </div>
          {this.state.userNameTouch.touched && <FormError message={userNameError} />}
          <div className="register-form-div">
            <label htmlFor="password" className="register-label">Password<span className="required">*</span></label>
            <input type="password" className="register-input" id="password" name="password" minLength="8" maxLength="72" required
              autoComplete="on" aria-required="true" aria-label="password" aria-invalid="true"
              onChange={e => {this.handleChange(e);this.handlePasswordTouch(e)}}/>
          </div>
          {this.state.passwordTouch.touched && <FormError message={passwordError} />}
          <p className="password-requirements"><span className="required">*</span>Password must contain at least one uppercase, lowercase, and number character</p>
          <div className="buttons-flex">
            <div className="signup-button-div">
              <button type="submit" className="signup-button" disabled={fullNameError || userNameError || passwordError}>Sign Up</button>
            </div>
            <div className="back-button-div">
              <button type="button" className="back-button" onClick={() => this.props.history.push('/')}>Back</button>
            </div>
          </div>
        </form>
      </section>
    )
  }
}