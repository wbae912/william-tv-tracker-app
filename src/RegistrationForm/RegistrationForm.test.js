import React from 'react';
import ReactDOM from 'react-dom';
import RegistrationForm from './RegistrationForm';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';

describe('RegistrationForm component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<MemoryRouter><RegistrationForm /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  
  it('renders UI as expected', () => {
    const tree = renderer
      .create(<MemoryRouter><RegistrationForm /></MemoryRouter>)
      .toJSON();
      expect(tree).toMatchSnapshot();
  })
})