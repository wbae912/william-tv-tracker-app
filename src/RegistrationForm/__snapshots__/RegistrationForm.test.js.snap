// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm component renders UI as expected 1`] = `
<section
  className="register-section"
>
  <div
    className="error-div"
    role="alert"
  />
  <h2
    className="register-h2"
  >
    Start Today!
  </h2>
  <form
    className="register-form"
    onSubmit={[Function]}
  >
    <div
      className="register-form-div"
    >
      <label
        className="register-label"
        htmlFor="full_name"
      >
        Full Name
      </label>
      <input
        className="register-input"
        id="full_name"
        name="full_name"
        onChange={[Function]}
        placeholder="Bob Smith"
        required={true}
        type="text"
      />
    </div>
    <div
      className="register-form-div"
    >
      <label
        className="register-label"
        htmlFor="user_name"
      >
        Username
      </label>
      <input
        className="register-input"
        id="user_name"
        name="user_name"
        onChange={[Function]}
        placeholder="username"
        required={true}
        type="text"
      />
    </div>
    <div
      className="register-form-div"
    >
      <label
        className="register-label"
        htmlFor="password"
      >
        Password
        <span
          className="required"
        >
          *
        </span>
      </label>
      <input
        autoComplete="on"
        className="register-input"
        id="password"
        maxLength="72"
        minLength="8"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
      />
    </div>
    <p
      className="password-requirements"
    >
      <span
        className="required"
      >
        *
      </span>
      Password must contain at least one uppercase, lowercase, and number character
    </p>
    <div
      className="buttons-flex"
    >
      <div
        className="signup-button-div"
      >
        <button
          className="signup-button"
          disabled="Please enter your full name"
          type="submit"
        >
          Sign Up
        </button>
      </div>
      <div
        className="back-button-div"
      >
        <button
          className="back-button"
          onClick={[Function]}
          type="button"
        >
          Back
        </button>
      </div>
    </div>
  </form>
</section>
`;
